src/Microsoft.AspNet.OData.Shared/Query/SkipTokenQueryOption.cs

```diff
@@ -0,0 +1,168 @@
+                context.RequestContainer);
+        }
+
+        /// <summary>
+        /// Gets the raw $skiptoken value.
+        /// </summary>
+        public string RawValue { get; private set; }
+
+        /// <summary>
+        /// Gets and sets the given <see cref="ODataQueryContext"/>.
+        /// </summary>
+        public ODataQueryContext Context { get; set; }
+
+        /// <summary>
+        /// Generates the nextlink value and consumes the skiptoken value.
+        /// </summary>
+        private SkipTokenHandler SkipTokenHandler { get; set; }
+
+        /// <summary>
+        /// Gets the value of the $skiptoken.
+        /// </summary>
+        public string Value
+        {
+            get
+            {
+                if (_value == null)
+                {
+                    _value = _queryOptionParser.ParseSkipToken();
+                }
+                return _value;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the SkipToken Query Validator.
+        /// </summary>
+        public SkipTokenQueryValidator Validator { get; set; }
+
+        /// <summary>
+        /// Gets or sets the query setting 
+        /// </summary>
+        public ODataQuerySettings QuerySettings { get; private set; }
+
+        /// <summary>
+        /// Gets or sets the list of OrderByNodes for the original query
+        /// </summary>
+        public IList<OrderByNode> OrderByNodes { get; private set; }
+
+        /// <summary>
+        /// Apply the $skiptoken query to the given IQueryable.
+        /// </summary>
+        /// <param name="query">The original <see cref="IQueryable"/>.</param>
+        /// <param name="querySettings">The query settings to use while applying this query option.</param>
+        /// <param name="orderByNodes">Information about the orderby query option.</param>
+        /// <returns>The new <see cref="IQueryable"/> after the skiptoken query has been applied to.</returns>
+        public virtual IQueryable<T> ApplyTo<T>(IQueryable<T> query, ODataQuerySettings querySettings, IList<OrderByNode> orderByNodes)
+        {
+            QuerySettings = querySettings;
+            OrderByNodes = orderByNodes;
+            return SkipTokenHandler.ApplyTo<T>(query, this) as IOrderedQueryable<T>;
+        }
+
+        /// <summary>
+        /// Apply the $skiptoken query to the given IQueryable.
+        /// </summary>
+        /// <param name="query">The original <see cref="IQueryable"/>.</param>
+        /// <param name="querySettings">The query settings to use while applying this query option.</param>
+        /// <param name="orderByNodes">Information about the orderby query option.</param>
+        /// <returns>The new <see cref="IQueryable"/> after the skiptoken query has been applied to.</returns>
+        public virtual IQueryable ApplyTo(IQueryable query, ODataQuerySettings querySettings, IList<OrderByNode> orderByNodes)
```

Comment:
>IList<OrderByNode> orderByNodes [](start = 94, length = 31)

Consider replacing orderByNodes with ODataQueryOptions, and provide a method on ODataQueryOptions for GetStableOrdering (that caches the result).